central/scannerdefinitions/handler/handler.go

```diff
@@ -535,96 +522,110 @@ func (h *httpHandler) post(w http.ResponseWriter, r *http.Request) {
+			}
+			log.Infof("v2 definitions: found supported bundle: %s", zipF.Name)
+			defs[offlineScannerV2DefsBlobName] = zipF
+		case features.ScannerV4.Enabled() && strings.HasPrefix(zipF.Name, scannerV4DefsPrefix):
+			ok, err := checkV4Manifest(zipF)
+			if err != nil {
+				// If we cannot check if the V4 definitions is supported, then we assume the
+				// bundle is not good, which is not entirely accurate but simplify error
+				// handling.
+				return fmt.Errorf("invalid v4 definition: %s: %w", zipF.Name, err)
+			}
+			if !ok {
+				// Ignore unsupported bundles.
+				log.Debugf("v4 definitions: ignoring unsupported bundle: %s", zipF.Name)
+				continue
 			}
-			log.Debugf("Successfully processed file: %s", zipF.Name)
-			count++
-		case strings.HasPrefix(zipF.Name, scannerV4DefsPrefix):
-			if err := h.handleScannerDefsFile(ctx, zipF, offlineScannerV4DefsBlobName); err != nil {
-				return errors.Wrap(err, "couldn't handle scanner-v4-defs sub file")
+			if prev, ok := defs[offlineScannerV4DefsBlobName]; ok {
+				return fmt.Errorf("v4 definitions: found more than one supported definitions: %s and %s",
+					prev.Name, zipF.Name)
 			}
-			log.Debugf("Successfully processed file: %s", zipF.Name)
-			count++
+			log.Infof("v4 definitions: found supported bundle: %s", zipF.Name)
+			defs[offlineScannerV4DefsBlobName] = zipF
 		default:
 			// Ignore any other files which may be in the ZIP.
 		}
 	}
-	// Just do a simple check for at least one valid file.
-	// Anything stricter may come with incompatibilities.
-	if count > 0 {
-		return nil
+
+	// Check bundle compatibility.
+	var incompatible bool
+	if _, ok := defs[offlineScannerV4DefsBlobName]; features.ScannerV4.Enabled() && !ok {
+		incompatible = true
+		log.Debugf("offline bundle compatibility check: missing V4 definitions (and V4 is enabled)")
+	}
+	if _, ok := defs[offlineScannerV2DefsBlobName]; !ok {
+		incompatible = true
+		log.Debugf("offline bundle compatibility check: missing V2 definitions")
+	}
+	if incompatible {
+		// If the expected definitions were not found, we assume the bundle is incompatible.
+		return fmt.Errorf("the uploaded "+
+			"bundle is incompatible with release version number: %s: "+
+			"please upload an offline bundle that supports this "+
+			"release, and consider using `roxctl scanner download-db`",
+			mainVersionNum)
+	}
+
+	var errList []error
+	for blobName := range defs {
+		zipF := defs[blobName]
+		if err := h.handleScannerDefsFile(ctx, zipF, blobName); err != nil {
+			err = fmt.Errorf("failed to load %s: %w", zipF.Name, err)
+			log.Error(err)
+			errList = append(errList, err)
+		}
+		log.Infof("Successfully loaded offline definition: %s", zipF.Name)
 	}
 
-	return errors.New("scanner defs file(s) not found in upload zip: incorrect zip?")
+	if len(errList) > 0 {
+		return fmt.Errorf("failed to load %d out of %d scanner defs file(s): %v",
+			len(errList), len(defs), errList)
```

Comment:
which formatting do you think is better: `errList` or `errors.Join(errList...)`?

Reply:
Slightly biased towards errors on a single line to favor machine parsing. But fine to move away from that.